builtin/aws/ecs/task.go

```diff
@@ -209,24 +212,152 @@ func (p *TaskLauncher) Config() (interface{}, error) {
+		Family:  aws.String("waypoint-runner"),
+	})
+	if err != nil {
+		return err
+	}
+	for _, taskArn := range tasks.TaskArns {
+		taskResp, err := ecsSvc.DescribeTasks(&ecs.DescribeTasksInput{
+			Cluster: aws.String(p.config.Cluster),
+			Tasks:   aws.StringSlice([]string{*taskArn}),
+			Include: aws.StringSlice([]string{"TAGS"}),
+		})
+		if err != nil {
+			return err
+		} else if len(taskResp.Tasks) != 1 {
+			return status.Errorf(codes.Internal, "there should be only 1 task, but there are %d", len(taskResp.Tasks))
+		}
+		for _, tag := range taskResp.Tasks[0].Tags {
+			if *tag.Key == "waypoint-odr-task-name" && *tag.Value == ti.Id {
+				log.Info("stopping ECS task", "task_id", ti.Id)
+				_, err := ecsSvc.StopTask(&ecs.StopTaskInput{
+					Cluster: aws.String(p.config.Cluster),
+					Reason:  aws.String("Waypoint ODR job is complete"),
+					Task:    taskArn,
+				})
+				if err != nil {
+					return err
+				}
+				return nil
+			}
+		}
+	}
+
+	// If we reach this point, then we could not find our task in ECS
+	// so there's nothing to clean up. Normally, ECS will automatically
+	// clean up the task.
 	return nil
 }
 
 // WatchTask implements TaskLauncher
 func (p *TaskLauncher) WatchTask(
 	ctx context.Context,
 	log hclog.Logger,
+	ui terminal.UI,
 	ti *TaskInfo,
 ) (*component.TaskResult, error) {
-	return nil, status.Errorf(codes.Unimplemented, "WatchTask not implemented")
+	sess, err := utils.GetSession(&utils.SessionConfig{
+		Region: p.config.Region,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// this channel will receive the status of the ECS task
+	taskStatusCh := make(chan string)
+	go taskStatus(ctx, time.Duration(5*time.Minute), &taskStatusCh, log, sess, p.config.Cluster, ti.Id)
+
+	var logStreamName string
+	// We wait 5 minutes for the log stream to be available
+	logStreamContext, cancel := context.WithTimeout(ctx, time.Minute*5)
+	cwl := cloudwatchlogs.New(sess)
+	// loop until task is ready
+	defer cancel()
+	for logStreamName == "" {
+		select {
+		case <-logStreamContext.Done():
+			log.Error("timeout waiting for log stream to become ready", terminal.WithErrorStyle())
+			return nil, err
+		case taskStatus := <-taskStatusCh:
+			if taskStatus == "RUNNING" {
+				// use the task prefix "waypoint-odr-task-<id>" to filter the log streams
+				logStreamsResp, err := cwl.DescribeLogStreams(&cloudwatchlogs.DescribeLogStreamsInput{
+					LogGroupName:        aws.String(p.config.LogGroup),
+					LogStreamNamePrefix: aws.String(ti.Id),
+				})
+				if err != nil {
+					log.Error("error describing log streams", "err", err)
+					ui.Output("Error describing log streams: %s", err)
+					return nil, err
+				}
+
+				for _, logStream := range logStreamsResp.LogStreams {
+					if strings.Contains(*logStream.LogStreamName, ti.Id) {
+						logStreamName = *logStream.LogStreamName
+					}
+				}
+			}
+		default:
+			continue
+		}
+	}
+
+	token := ""
+	var resp *cloudwatchlogs.GetLogEventsOutput
+	for {
```

Comment:
Is it possible to refactor this to also have a `for { select { ... } }` pattern like the previous loop? I feel like with unbound loops those are really nice for readability 🤔

Reply:
@briancain added! But do you also mean to add another timeout context for the `select` to read from?